### PR TITLE
python packaging improvements

### PR DESCRIPTION
**Pull request by bpkroth**
# Pull Request

## Title

A collection of changes for python packaging improvements.

---

## Description

- [x] Reorganizes the top level directory
- [x] Changes the `import` lines accordingly
- [x] Replaces `requirements.txt` and `setup.py` with `pyproject.toml`
- [x] Adds `pre-commit` to the `dev` dependencies.

See Also:

- #39
- #58

---

## Type of Change

- Refactor

---

## Additional Notes (optional)

Should be merged after #63

### CODE REVIEW

File: .github/workflows/pythonapp.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
       matrix:
         # TODO: We will keep this simple for now. We will add more versions later.
         os: [ubuntu-22.04]
-        python-version: ['3.10']
+        python-version: ['3.10', '3.12']
```
Comment (bpkroth) [inline .github/workflows/pythonapp.yml:28]: Move to separate PR.

File: .github/workflows/pythonapp.yml [hunk 2]
```diff
@@ -55,7 +55,7 @@ jobs:
         else
           echo "All files have the required license headers."
         fi
-    - name: Install basic dependencies # TODO: separate out "test" dependencies in setup.py
+    - name: Install test dependencies
```
Comment (bpkroth) [inline .github/workflows/pythonapp.yml:58]: Minor comment fixup - separate PR.

File: CONTRIBUTING.md
Comment (bpkroth) [inline CONTRIBUTING.md:1]: Should be in #63

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+Repository = "https://github.com/microsoft/vasim"
+Issues = "https://github.com/microsoft/vasim/issues"
+
+# Tell setuptools where to find the package.
+[tool.setuptools]
+package-dir = {"" = "src"}
+
+[tool.pytest.ini_options]
+minversion = "6.0"
+addopts = "-s -n auto"
```
Comment (bpkroth) [inline pyproject.toml:66]: Move the `xdist` options to a separate PR.

Comment (bpkroth) [reply pyproject.toml]: See Also #60

File: .github/workflows/pythonapp.yml (hunk 2 above)
Comment (bpkroth) [reply .github/workflows/pythonapp.yml]: #65

**Comment by apavlen** [comment]
There was previously a known compatibility issue between sktime and numpy, where sktime could fail during prediction without providing a clear error message.And it was the reason to pin sktime to specific version. I don't think we have a unit test right now to catch it, so need to manually test before merging or add a unit test

File: examples/using_vasim.ipynb [hunk 4]
```diff
@@ -332,7 +332,7 @@
     }
    ],
    "source": [
-    "from simulator.InMemorySimulator import InMemoryRunnerSimulator\n",
+    "from vasim.simulator.InMemorySimulator import InMemoryRunnerSimulator\n",
```
Comment (ksaur) [inline examples/using_vasim.ipynb:335]: Can you please also run this notebook and make sure output all looks the same?  (No need to check that part in because output shouldn't change!)  
meant to pin that to a particular version/release but i didn't do that yet

I just want to avoid surprises here since that is currently our documentation :D

**Review by apavlen — APPROVED** [review]
LGTM! Thank you!

File: examples/using_vasim.ipynb (hunk 4 above)
Comment (bpkroth) [reply examples/using_vasim.ipynb:335]: Done!

**Comment by bpkroth** [comment]
> There was previously a known compatibility issue between sktime and numpy, where sktime could fail during prediction without providing a clear error message.And it was the reason to pin sktime to specific version. I don't think we have a unit test right now to catch it, so need to manually test before merging or add a unit test

Should probably have an issue made for this.

I also asked about this in #3 I think.
